### DISCUSSION TITLE
Building FlatBuffers in an embedded environment.

### DISCUSSION THREAD

**Discussion by randyK-HH** [dvidelabs/flatcc#195] [Q&A] [answered]
I would like to use FlatBuffers in an embedded system and am a little unsure of what the best way to do this is. 

I have built the flatcc compiler on a desktop computer (Mac) with no problem and have been able to compile and test the (sample) monster schema properly there. So far so good! Now I want to build the flatcc libraries for an ARM Cortex M0 processor so that I can use them in an embedded application.

Any general advice on how to do this?

**Comment by mikkelfj** [comment, marked as answer]
So, flatcc is designed for this use case, but I rely on users to test various embedded devices.
I believe systems similar to the one you mention have used flatcc successfully.

Essentiall you just compile the runtime parts for your target system either as a library, or add the few runtime C files to your project along with your application. I will not go into how you can get CMake to cross compile and you can also just compile with a shell script if so inclined (I often do that for testing small projects). You may want to look at the compiler flags set in the CMake project because gcc can be a bit loud on some warnings, but you do not have to use the CMake system that flatcc builds with.

The main problems typically arise from not using a full standard library and in particular when malloc and friends are not available. FreeRTOS systems have compiled and most of these issues have been ironed out, but the `<flatcc/portable>` sub library can be updated if you run into specific problems, and allocation functions can be redefined via macros.

Another problem arises for extremely RAM limited use cases. Here the builders internal allocation system may have to be rewritten via a pluggable allocator so it can allocate from a fixed memory block or similar. This is possible but a bit involved.

**Reply by randyK-HH** [reply]
Thanks @mikkelfj for the response! Some follow up questions:

1. Exactly which files make up the "runtime parts" ? 
2. You mentioned problem from RAM limited cases. How much RAM is considered to be a bare minimum? And what sorts of problems arise from a lack of RAM?

**Reply by mikkelfj** [reply]
**Runtime parts**

* https://github.com/dvidelabs/flatcc/tree/master/src/runtime
* https://github.com/dvidelabs/flatcc#cross-compilation

You don't need all, depending on context - it is  mentioned in doc, but builder.c is the bare minimum for building buffers (and refmap.c if you use cloning methods). If you only read buffers without building, verifying, or processing JSON, you do not need runtime files, except for the include directory which is always needed.

**RAM limits**

I don't have exact numbers, but I would expect 2-4K to suffice for most small buffer projects - you can get significantly below that if you work hard enough and if your data is simple enough.

Unlike other languages, flatcc attempts to be userfriendly by providing a stack (several stacks actually). It allows you to create objects while a containing object is still open, e.g. creating a string and adding it to an open table, or appending data to a vector that is added to a table and so forth. Depending on how this is done, significant amounts of stack memory may be required temporarily. I won't go into all the details here, but if you run into specific problems, I can fill in the gaps. If you are very RAM limited, you can benchmark the different types of memory allocated by hooking into the default allocator, then replace it with a custom allocator that makes sure that just about enough is allocated initially for each type of allocation - and error hard if reallocation is required because that would not be possible if allocating memory statically at startup.

A related problem is that flatcc user code that build buffers generally do not check for error conditions because only two things can go wrong: wrong use of API, and out of memory, and because it is exceedingly tedious to do so. If you have very constrained memory you want to add some logging to catch such cases rather than pollute user code.

* https://github.com/dvidelabs/flatcc#custom-allocation

**Reply by randyK-HH** [reply]
I think I am still missing some context here. Thanks for hanging in there with me @mikkelfj.

My understanding of how this works is that I build the flatcc compiler on my target system (currently my desktop/Mac) and once built, I can then compile a custom schema.

Compiling the monster schema (monster.fbs) created a handful of files that I can them import into a different C project.

Files created from the schema compile were:

flatbuffers_common_builder.h
flatbuffers_common_reader.h
monster_builder.h
monster_reader.h
monster_verifier.h

AND, library files:
libflatcc.a (et al)

Since the .h files are standard C includes, I would expect I can easily import them into an embedded project and use them there. 

Since the libraries were compiled on my original target system, I would expect that I can NOT import them into an embedded project and use them there.

Any disagreement with those statements?


So! What I am really trying to do here (I think) is build the library files for my embedded target.

First though, is libflatcc.a (et al) agnostic to the schema?

If so, that seems to tell me that compiling the runtime files plus the includes will generate the libflatcc libraries on my embedded platform. 

Do you agree? (If not, where did I diverge?)

Also, when we say “includes” I assume you are referring to all the include files found in the 
/include/flatcc/ directory.

One thing you said which may be confusing me is that to read buffers, you do not need runtime files. However, how can we read the buffers if we don’t first have the flatcc libraries compiled for our platform?

**Reply by mikkelfj** [reply]
Yes you almost got it right. Note that there is libflatccrt.a and libflatcc.a. The first is the runtime library, the second is schema compiler, which also includes the runtime library. You do not want libflatcc.a because it is larger and potentially harder to port. libflatccrt.a is just the files I mentioned in my previous post.

And yes, includes are those in include/flatcc. You would want to add that path to your compilers -I include path.

The runtime files are only needed for a limited number of purposes, most notably building buffers.
All platform abstractions happen in header files. Generated code use macro expansion to produce a lot of inline functions that ultimately rely on the flatcc header files, and sometimes also the runtime library (for builders, JSON parsers, and JSON printers).

You don't really need to build a library for your target. If you just add builder.c (EDIT: and also emitter.c) you will go a long way, and as stated, for reading buffers it is not needed because the generated inline functions are sufficient. Flatbuffers are fast and reading a table field is only a few instructions and no dynamic memory management.

**Reply by mikkelfj** [reply]
I should add: just go ahead and try instead of second guessing yourself. It isn't that hard if you have a compiler for your target installed. But of course test the setup first on your host to rule out simple mistakes. That is, without relying on the official flatcc build products - except for the flatcc executable that you need to generate code.

**Reply by randyK-HH** [reply]
Is indeed the libflatcc.a (or libflatccrt.a) library agnostic to the schema?

I did, however, do as you said and imported builder.c and the include files into a small sample project (on my ARM target). The compiler errors out because function 'aligned_alloc' is not recognized. Any suggestions or workarounds about how to get past this?

```
Building file: ../builder.c
Invoking: Cross ARM C Compiler
../builder.c: In function 'flatcc_builder_finalize_aligned_buffer':
../builder.c:1995:5: error: implicit declaration of function 'aligned_alloc' [-Werror=implicit-function-declaration]
```

Compiler version I am using is: GNU ISO C11

**Reply by mikkelfj** [reply]
Yes, the library is schema agnostic. Everything that is not generated is schema agnostic, but you should try to match the version of flatcc with the version of library and headers.

`aligned_alloc` is a typical problem on new platforms because there is no standard for doing so except for C11 which isn't widely supported, or not available by default.

The implementation is done here:

https://github.com/dvidelabs/flatcc/blob/master/include/flatcc/portable/paligned_alloc.h

It attempts to detect the platform and use available means to implement aligned allocation, or fall back a custom implementation based on malloc.

You can define `#define PORTABLE_DEBUG_ALIGNED_ALLOC` using `-DPORTABLE_DEBUG_ALIGNED_ALLOC` as compiler argument to get debug information on how aligned alloc fails. This is mentioned on the header of paligned_alloc.h

The fix is to update paligned_alloc.h to also handle your platform, or make sure that you provide some defines necessary for paligned_alloc.h to do the right thing. For example you may have to define `_ISOC11_SOURCE` or `_POSIX_C_SOURCE ` as dicussed in the header of paligned_alloc.h. Or may also suffice to compile with -std=c11 if your project allows for that. It also depends on your glibc version, if you are using that C runtime library. Some ARM users use newlib which should work with flatcc.

**Reply by mikkelfj** [reply]
I missed your last comment: `Compiler version I am using is: GNU ISO C11`. Defining PORTABLE_DEBUG_ALIGNED_ALLOC should give you more information. When you have that, let me know.

**Reply by mikkelfj** [reply]
You could also try to define `-DPORTABLE_C11_ALIGNED_ALLOC=0` This should disable C11 aligned_alloc and use a fallback solution in case your library does not really support C11.

**Reply by randyK-HH** [reply]
When I #define PORTABLE_DEBUG_ALIGNED_ALLOC=1 as a compiler argument, error messages I get are:

```
Building file: ../builder.c
Invoking: Cross ARM C Compiler
In file included from C:\Workspaces\freertos_retarget/flatcc/portable/pstdalign.h:160:0,
                 from ../flatcc/flatcc_flatbuffers.h:24,
                 from ../flatcc/flatcc_builder.h:68,
                 from ../builder.c:19:
C:\Workspaces\freertos_retarget/flatcc/portable/paligned_alloc.h:111:2: error: #error "DEBUG: C11_ALIGNED_ALLOC configured"
 #error "DEBUG: C11_ALIGNED_ALLOC configured"
  ^
In file included from C:\Workspaces\freertos_retarget/flatcc/flatcc_alloc.h:72:0,
                 from ../flatcc/flatcc_flatbuffers.h:26,
                 from ../flatcc/flatcc_builder.h:68,
                 from ../builder.c:19:
C:\Workspaces\freertos_retarget/flatcc/portable/paligned_alloc.h:111:2: error: #error "DEBUG: C11_ALIGNED_ALLOC configured"
 #error "DEBUG: C11_ALIGNED_ALLOC configured"
```

When I #define PORTABLE_C11_ALIGNED_ALLOC=0 (and remove #define PORTABLE_DEBUG_ALIGNED_ALLOC), the project does appear to compile successfully. That seems good! Next step is to try it out.

**Reply by randyK-HH** [reply]
One minor snag I am encountering now is from the linker. When I call: `flatcc_builder_init(&builder);` from my main routine I get:

```
Building file: ../builder.c
Invoking: Cross ARM C Compiler
Finished building: ../builder.c
 
Building file: ../main.c
Invoking: Cross ARM C Compiler
Finished building: ../main.c
 
Building target: freertos_retarget.elf
Invoking: Cross ARM C Linker
./builder.o: In function `flatcc_builder_init':
C:\Workspaces\<PATH REMOVED FOR BREVITY>/../builder.c:425: undefined reference to `flatcc_emitter'
collect2.exe: error: ld returned 1 exit status
make: *** [freertos_retarget.elf] Error 1
```

I can see flatcc_emitter is a function declaration (ie: no body) at the bottom of flatcc_emitter.h. That seems suspicious and seems like it might be part of the issue.

I also see there is a struct defined in flatcc_emitter.h with the name flatcc_emitter.

**Reply by mikkelfj** [reply]
It is because I forgot that you also need the emitter.c file when building buffers. It is pluggable module that controls how the output buffer is allocated and emitter.c provides the default implemenation.

**Reply by randyK-HH** [reply]
That was indeed the issue! Compiles/links successfully. Thank you so much for the help!

Upon running, I noticed that I had to increase my heap size to prevent a crash which I believe was associated with memory allocation. I boosted it from 0x0600 (1536) to 0x2000 (8192). Nothing magical about the 0x2000, I just arbitrarily picked a bigger value. 

Previously, you mentioned 2-4K was sufficient for small projects. Does this new heap value seem consistent with your expectation?

**Reply by mikkelfj** [reply]
> Previously, you mentioned 2-4K was sufficient for small projects. Does this new heap value seem consistent with your expectation?

You can try to reduce the default page size of the emitter. It appears to be 3K.

The change the page size define e.g.: -DFLATCC_EMITTER_PAGE_SIZE=1024
see also
https://github.com/dvidelabs/flatcc/blob/master/include/flatcc/flatcc_emitter.h

And you make sure you only use direct_buffer when reading the buffer to avoid allocating a congtigous buffer for the result: flatcc_builder_get_direct_buffer
However!!! it only works for small buffers about half a page size. Otherwise it will return null, and multiple pages will have been allocated. The emitter starts in the middle of a page in order to extend in both directions for reasons I will not go into here. But you can ask again if you want to optimize this.

The above pertains to the emitter which builds up the buffer in reusable pages. In addition there is a number of small allocations for the builder stack - even if they are initially small, the system allocator may also reserve extra memory for each allocated size, so it is hard to tell exactly what the memory requirements are for a given system.

You can look at flatcc_builder_default_alloc in builder.c - each hint option provides an initial allocation size. You can copy and modify the allocator and give to the builders extended init function. Then you can debug how much allocation is needed and modify the allocator accordingly.
